File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        // 컨테이너 벨트 는 큐
        // 보조 컨테이너 베트는 스택
        Queue<Integer> belt = new LinkedList<>();
        Stack<Integer> subBelt = new Stack<>();
        
        for(int i = 1; i <= order.length ; i++){
            belt.add(i);
        }

        for(int o : order){
            if(!subBelt.isEmpty() && subBelt.peek() == o){
                answer += 1;
                subBelt.pop();
                continue;
            }

            while(!belt.isEmpty() && belt.peek() != o){
                subBelt.add(belt.poll());
            }

            if(!belt.isEmpty() && belt.peek() == o){
                belt.poll();
                answer += 1;
            } else {
                break;
            }
        }
        return answer;
    }
}

File: 프로그래머스/2/131704. 택배상자/택배상자.kt
import java.util.*

class Solution {
fun solution(order: IntArray): Int {
    var answer: Int = 0
    // 3/1
    // 4/2
    //2/3
    //1/4
    //5/5
    val pq = PriorityQueue<Pair<Int,Int>>(compareBy { it.second })
    order.forEachIndexed { index, o ->
        pq.add(Pair(index+1, o))
    }

    val subConveyor = Stack<Int>()
    var current = 1
    while(pq.isNotEmpty()){
        val (realOrder, _ ) = pq.poll()
        var flag = false

        while(subConveyor.isNotEmpty()){
            if(subConveyor.peek() == current){
                // println("peek current from sub : $current")
                subConveyor.pop()
                current += 1
//                flag = true
            } else {
                break
            }
        }

        if(current == realOrder) {
            // println("peek current from pri : $current")
            current+= 1
//            flag = true
        } else {
            subConveyor.push(realOrder)
            // println("push to sub : $realOrder")
        }

    }

    while(subConveyor.isNotEmpty()){
        if(subConveyor.peek() == current){
            // println("peek current from sub : $current")
            subConveyor.pop()
            current += 1
//                flag = true
        } else {
            break
        }
    }
    
    return current-1
}

}